File: app/src/main/java/org/tvheadend/tvhclient/ui/features/dvr/RecordingUtils.java
package org.tvheadend.tvhclient.ui.features.dvr;

import android.content.Context;
import android.os.Bundle;

import com.afollestad.materialdialogs.MaterialDialog;

import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.domain.entity.Channel;
import org.tvheadend.tvhclient.domain.entity.ServerProfile;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.CopyOnWriteArrayList;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;
import timber.log.Timber;

public class RecordingUtils {

    public static String getPriorityName(Context context,  int priority) {
        String[] priorityNames = context.getResources().getStringArray(R.array.dvr_priority_names);
        if (priority >= 0 && priority <= 4) {
            return priorityNames[priority];
        } else if (priority == 6) {
            return priorityNames[5];
        } else {
            return "";
        }
    }

    public static String getDateStringFromTimeInMillis(long milliSeconds) {
        SimpleDateFormat sdf = new SimpleDateFormat("dd:MM", Locale.US);
        return sdf.format(milliSeconds);
    }

    public static String getTimeStringFromTimeInMillis(long milliSeconds) {
        SimpleDateFormat sdf = new SimpleDateFormat("HH:mm", Locale.US);
        return sdf.format(milliSeconds);
    }

    public static String getSelectedDaysOfWeekText(Context context, int daysOfWeek) {
        String[] daysOfWeekList = context.getResources().getStringArray(R.array.day_short_names);
        StringBuilder text = new StringBuilder();
        for (int i = 0; i < 7; i++) {
            String s = (((daysOfWeek >> i) & 1) == 1) ? daysOfWeekList[i] : "";
            if (text.length() > 0 && s.length() > 0) {
                text.append(", ");
            }
            text.append(s);
        }
        return text.toString();
    }

    public static void handleDayOfWeekSelection(Context context, int daysOfWeek, RecordingConfigSelectedListener callback) {
        // Get the selected indices by storing the bits with 1 positions in a list
        // This list then needs to be converted to an Integer[] because the
        // material dialog requires this
        List<Integer> list = new ArrayList<>();
        for (int i = 0; i < 7; i++) {
            int value = (daysOfWeek >> i) & 1;
            if (value == 1) {
                list.add(i);
            }
        }
        Integer[] selectedIndices = new Integer[list.size()];
        for (int i = 0; i < selectedIndices.length; i++) {
            selectedIndices[i] = list.get(i);
        }
        new MaterialDialog.Builder(context)
                .title(R.string.days_of_week)
                .items(R.array.day_long_names)
                .itemsCallbackMultiChoice(selectedIndices, (dialog, which, text) -> {
                    int selectedDays = 0;
                    for (Integer i : which) {
                        selectedDays += (1 << i);
                    }
                    if (callback != null) {
                        callback.onDaysSelected(selectedDays);
                    }
                    return true;
                })
                .positiveText(R.string.select)
                .show();
    }

    public static void handleChannelListSelection(Context context, List<Channel> channelList, boolean showAllChannelsListEntry, RecordingConfigSelectedListener callback) {
        // Fill the channel tag adapter with the available channel tags
        CopyOnWriteArrayList<Channel> channels = new CopyOnWriteArrayList<>(channelList);

        // Add the default channel (all channels)
        // to the list after it has been sorted
        if (showAllChannelsListEntry) {
            Channel channel = new Channel();
            channel.setId(0);
            channel.setName(context.getString(R.string.all_channels));
            channels.add(0, channel);
        }

        final ChannelListSelectionAdapter channelListSelectionAdapter = new ChannelListSelectionAdapter(context, channels);
        // Show the dialog that shows all available channel tags. When the
        // user has selected a tag, restart the loader to loadRecordingById the updated channel list
        final MaterialDialog dialog = new MaterialDialog.Builder(context)
                .title(R.string.tags)
                .adapter(channelListSelectionAdapter, null)
                .build();
        // Set the callback to handle clicks. This needs to be done after the
        // dialog creation so that the inner method has access to the dialog variable
        channelListSelectionAdapter.setCallback(channel -> {
            if (callback != null) {
                callback.onChannelSelected(channel);
            }
            if (dialog != null) {
                dialog.dismiss();
            }
        });
        dialog.show();
    }

    public static void handlePrioritySelection(Context context, int selectedPriority, RecordingConfigSelectedListener callback) {
        Timber.d("Selected priority is " + (selectedPriority == 6 ? 5 : selectedPriority));
        String[] priorityNames = context.getResources().getStringArray(R.array.dvr_priority_names);
        new MaterialDialog.Builder(context)
                .title(R.string.select_priority)
                .items(priorityNames)
                .itemsCallbackSingleChoice(selectedPriority == 6 ? 5 : selectedPriority, (dialog, view, which, text) -> {
                    if (callback != null) {
                        Timber.d("New selected priority is " + (which == 5 ? 6 : which));
                        callback.onPrioritySelected(which == 5 ? 6 : which);
                    }
                    return true;
                })
                .show();
    }

    public static void handleRecordingProfileSelection(Context context, String[] recordingProfilesList, int selectedProfile, RecordingConfigSelectedListener callback) {
        new MaterialDialog.Builder(context)
                .title(R.string.select_dvr_config)
                .items(recordingProfilesList)
                .itemsCallbackSingleChoice(selectedProfile, (dialog, view, which, text) -> {
                    if (callback != null) {
                        callback.onProfileSelected(which);
                    }
                    return true;
                })
                .show();
    }

    public static int getSelectedProfileId(ServerProfile profile, String[] recordingProfilesList) {
        if (profile != null) {
            for (int i = 0; i < recordingProfilesList.length; i++) {
                if (recordingProfilesList[i].equals(profile.getName())) {
                    return i;
                }
            }
        }
        return 0;
    }

    public static void handleDateSelection(AppCompatActivity activity, long milliSeconds, Fragment callback, String tag) {
        DialogFragment newFragment = new DatePickerFragment();
        Bundle bundle = new Bundle();
        bundle.putLong("milliSeconds", milliSeconds);
        newFragment.setArguments(bundle);
        newFragment.setTargetFragment(callback, 1);
        newFragment.show(activity.getSupportFragmentManager(), tag);
    }

    public static void handleTimeSelection(AppCompatActivity activity, long milliSeconds, Fragment callback, String tag) {
        DialogFragment newFragment = new TimePickerFragment();
        Bundle bundle = new Bundle();
        bundle.putLong("milliSeconds", milliSeconds);
        newFragment.setArguments(bundle);
        newFragment.setTargetFragment(callback, 1);
        newFragment.show(activity.getSupportFragmentManager(), tag);
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/dvr/RecordingUtils.kt
package org.tvheadend.tvhclient.ui.features.dvr

import android.content.Context
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.list.customListAdapter
import com.afollestad.materialdialogs.list.listItemsMultiChoice
import com.afollestad.materialdialogs.list.listItemsSingleChoice
import org.tvheadend.data.entity.Channel
import org.tvheadend.data.entity.ServerProfile
import org.tvheadend.tvhclient.R
import timber.log.Timber
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList

fun getPriorityName(context: Context, priority: Int): String {
    val priorityNames = context.resources.getStringArray(R.array.dvr_priority_names)
    return when (priority) {
        in 0..4 -> priorityNames[priority]
        6 -> priorityNames[5]
        else -> ""
    }
}

fun getDateStringFromTimeInMillis(milliSeconds: Long): String {
    val sdf = SimpleDateFormat("dd.MM", Locale.US)
    return sdf.format(milliSeconds)
}

fun getTimeStringFromTimeInMillis(milliSeconds: Long): String {
    val sdf = SimpleDateFormat("HH:mm", Locale.US)
    return sdf.format(milliSeconds)
}

fun getSelectedDaysOfWeekText(context: Context, daysOfWeek: Int): String {
    val daysOfWeekList = context.resources.getStringArray(R.array.day_short_names)
    val text = StringBuilder()
    for (i in 0..6) {
        val s = if (daysOfWeek shr i and 1 == 1) daysOfWeekList[i] else ""
        if (text.isNotEmpty() && s.isNotEmpty()) {
            text.append(", ")
        }
        text.append(s)
    }
    return text.toString()
}

fun handleDayOfWeekSelection(context: Context, daysOfWeek: Int, callback: RecordingConfigSelectedListener?) {
    // Get the selected indices by storing the bits with 1 positions in a list
    // This list then needs to be converted to an Integer[] because the
    // material dialog requires this
    val list = ArrayList<Int>()
    for (i in 0..6) {
        val value = daysOfWeek shr i and 1
        if (value == 1) {
            list.add(i)
        }
    }
    val selectedIndices = arrayOfNulls<Int>(list.size)
    for (i in selectedIndices.indices) {
        selectedIndices[i] = list[i]
    }
    MaterialDialog(context).show {
        title(R.string.days_of_week)
        positiveButton(R.string.select)
        listItemsMultiChoice(R.array.day_long_names) { _, index, _ ->
            var selectedDays = 0
            for (i in index) {
                selectedDays += 1 shl i
            }
            callback?.onDaysSelected(selectedDays)
        }
    }
}

fun handleChannelListSelection(context: Context, channelList: List<Channel>, showAllChannelsListEntry: Boolean, callback: RecordingConfigSelectedListener?) {
    // Fill the channel tag adapter with the available channel tags
    val channels = CopyOnWriteArrayList(channelList)

    // Add the default channel (all channels)
    // to the list after it has been sorted
    if (showAllChannelsListEntry) {
        val channel = Channel()
        channel.id = 0
        channel.name = context.getString(R.string.all_channels)
        channels.add(0, channel)
    }

    val channelListSelectionAdapter = ChannelListSelectionAdapter(context, channels)
    // Show the dialog that shows all available channel tags. When the
    // user has selected a tag, restart the loader to loadRecordingById the updated channel list
    val dialog: MaterialDialog = MaterialDialog(context).show {
        title(R.string.tags)
        customListAdapter(channelListSelectionAdapter)
    }

    // Set the callback to handle clicks. This needs to be done after the
    // dialog creation so that the inner method has access to the dialog variable
    channelListSelectionAdapter.setCallback(object : ChannelListSelectionAdapter.Callback {
        override fun onItemClicked(channel: Channel) {
            callback?.onChannelSelected(channel)
            dialog.dismiss()
        }
    })
}

fun handlePrioritySelection(context: Context, selectedPriority: Int, callback: RecordingConfigSelectedListener?) {
    Timber.d("Selected priority is ${if (selectedPriority == 6) 5 else selectedPriority}")
    MaterialDialog(context).show {
        title(R.string.select_priority)
        listItemsSingleChoice(R.array.dvr_priority_names, initialSelection = if (selectedPriority == 6) 5 else selectedPriority) { _, index, _ ->
            Timber.d("New selected priority is ${if (index == 5) 6 else index}")
            callback?.onPrioritySelected(if (index == 5) 6 else index)
        }
    }
}

fun handleRecordingProfileSelection(context: Context, recordingProfilesList: Array<String>, selectedProfile: Int, callback: RecordingConfigSelectedListener?) {
    MaterialDialog(context).show {
        title(R.string.select_dvr_config)
        listItemsSingleChoice(items = recordingProfilesList.toList(), initialSelection = selectedProfile) { _, index, _ ->
            callback?.onProfileSelected(index)
        }
    }
}

fun getSelectedProfileId(profile: ServerProfile?, recordingProfilesList: Array<String>): Int {
    if (profile != null) {
        for (i in recordingProfilesList.indices) {
            if (recordingProfilesList[i] == profile.name) {
                return i
            }
        }
    }
    return 0
}

fun handleDateSelection(activity: FragmentActivity?, milliSeconds: Long, callback: Fragment, tag: String) {
    activity?.let {
        val newFragment = DatePickerFragment()
        val bundle = Bundle()
        bundle.putLong("milliSeconds", milliSeconds)
        newFragment.arguments = bundle
        newFragment.setTargetFragment(callback, 1)
        newFragment.show(activity.supportFragmentManager, tag)
    }
}

fun handleTimeSelection(activity: FragmentActivity?, milliSeconds: Long, callback: Fragment, tag: String) {
    activity?.let {
        val newFragment = TimePickerFragment()
        val bundle = Bundle()
        bundle.putLong("milliSeconds", milliSeconds)
        newFragment.arguments = bundle
        newFragment.setTargetFragment(callback, 1)
        newFragment.show(activity.supportFragmentManager, tag)
    }
}
